### ISSUE TITLE
Can't index when open project.

### ISSUE THREAD

**Issue by larrylau** [casteng/i-pascal#44]
I got this error. how can I do?
```
Error while indexing D:\fpc\3.0.4\units\i386-win32\rtl-objpas\fmtbcd.ppu
To reindex this file IDEA has to be restarted

com.google.common.util.concurrent.ExecutionError: java.lang.AssertionError: Reentrant indexing
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2048)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3973)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4957)
	at com.siberika.idea.pascal.lang.compiled.PPUDecompilerCache.getContents(PPUDecompilerCache.java:165)
	at com.siberika.idea.pascal.lang.compiled.PPUDecompilerCache.decompile(PPUDecompilerCache.java:67)
	at com.siberika.idea.pascal.lang.compiled.PPUFileDecompiler.decompileText(PPUFileDecompiler.java:34)
	at com.siberika.idea.pascal.lang.compiled.PPUFileDecompiler.decompile(PPUFileDecompiler.java:26)
	at com.intellij.openapi.fileEditor.impl.LoadTextUtil.loadText(LoadTextUtil.java:492)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.getDocument(FileDocumentManagerImpl.java:187)
	at com.intellij.psi.AbstractFileViewProvider.getDocument(AbstractFileViewProvider.java:197)
	at com.intellij.psi.AbstractFileViewProvider$VirtualFileContent.getText(AbstractFileViewProvider.java:479)
	at com.intellij.psi.AbstractFileViewProvider.getContents(AbstractFileViewProvider.java:177)
	at com.intellij.psi.impl.source.PsiFileImpl.loadTreeElement(PsiFileImpl.java:208)
	at com.intellij.psi.impl.source.PsiFileImpl.calcTreeElement(PsiFileImpl.java:712)
	at com.intellij.psi.impl.source.PsiFileImpl.calcStubTree(PsiFileImpl.java:940)
	at com.siberika.idea.pascal.lang.compiled.PascalCompiledStubBuilder.buildFileStub(PascalCompiledStubBuilder.java:35)
	at com.intellij.psi.impl.compiled.ClassFileStubBuilder.buildStubTree(ClassFileStubBuilder.java:45)
	at com.intellij.psi.impl.compiled.ClassFileStubBuilder.buildStubTree(ClassFileStubBuilder.java:25)
	at com.intellij.psi.stubs.StubTreeBuilder.buildStubTree(StubTreeBuilder.java:61)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.lambda$computeValue$0(StubUpdatingIndex.java:130)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:932)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:57)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.computeValue(StubUpdatingIndex.java:112)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.computeValue(StubUpdatingIndex.java:108)
	at com.intellij.util.indexing.SingleEntryIndexer.map(SingleEntryIndexer.java:40)
	at com.intellij.util.indexing.SingleEntryIndexer.map(SingleEntryIndexer.java:30)
	at com.intellij.util.indexing.impl.MapReduceIndex.mapInput(MapReduceIndex.java:312)
	at com.intellij.util.indexing.VfsAwareMapReduceIndex.mapInput(VfsAwareMapReduceIndex.java:109)
	at com.intellij.util.indexing.impl.MapReduceIndex.calculateUpdateData(MapReduceIndex.java:268)
	at com.intellij.util.indexing.impl.MapReduceIndex.update(MapReduceIndex.java:250)
	at com.intellij.util.indexing.FileBasedIndexImpl.updateSingleIndex(FileBasedIndexImpl.java:1673)
	at com.intellij.util.indexing.FileBasedIndexImpl.lambda$doIndexFileContent$28(FileBasedIndexImpl.java:1603)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.freezeFileTypeTemporarilyIn(FileTypeManagerImpl.java:656)
	at com.intellij.util.indexing.FileBasedIndexImpl.doIndexFileContent(FileBasedIndexImpl.java:1562)
	at com.intellij.util.indexing.FileBasedIndexImpl.indexFileContent(FileBasedIndexImpl.java:1549)
	at com.intellij.util.indexing.UnindexedFilesUpdater.lambda$indexFiles$0(UnindexedFilesUpdater.java:109)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.lambda$run$0(CacheUpdateRunner.java:219)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1106)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.lambda$run$1(CacheUpdateRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:169)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:156)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.run(CacheUpdateRunner.java:232)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:294)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError: Reentrant indexing
	at com.intellij.util.indexing.FileBasedIndexImpl.updateSingleIndex(FileBasedIndexImpl.java:1667)
	at com.intellij.util.indexing.FileBasedIndexImpl.doInvalidateIndicesForFile(FileBasedIndexImpl.java:1799)
	at com.intellij.util.indexing.FileBasedIndexImpl.access$2600(FileBasedIndexImpl.java:99)
	at com.intellij.util.indexing.FileBasedIndexImpl$ChangedFilesCollector.lambda$null$1(FileBasedIndexImpl.java:2036)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:197)
	at com.intellij.util.indexing.FileBasedIndexImpl$ChangedFilesCollector.lambda$null$2(FileBasedIndexImpl.java:2030)
	at com.intellij.util.ConcurrencyUtil.withLock(ConcurrencyUtil.java:229)
	at com.intellij.util.indexing.FileBasedIndexImpl$ChangedFilesCollector.lambda$processFilesInReadAction$3(FileBasedIndexImpl.java:2028)
	at com.intellij.util.indexing.VfsEventsMerger.processChanges(VfsEventsMerger.java:95)
	at com.intellij.util.indexing.FileBasedIndexImpl$ChangedFilesCollector.processFilesInReadAction(FileBasedIndexImpl.java:2027)
	at com.intellij.util.indexing.FileBasedIndexImpl$ChangedFilesCollector.ensureUpToDate(FileBasedIndexImpl.java:1980)
	at com.intellij.util.indexing.FileBasedIndexImpl.ensureUpToDate(FileBasedIndexImpl.java:710)
	at com.intellij.util.indexing.FileBasedIndexImpl.processExceptions(FileBasedIndexImpl.java:892)
	at com.intellij.util.indexing.FileBasedIndexImpl.processValueIterator(FileBasedIndexImpl.java:968)
	at com.intellij.util.indexing.FileBasedIndexImpl.processValuesInScope(FileBasedIndexImpl.java:939)
	at com.intellij.util.indexing.FileBasedIndexImpl.getContainingFiles(FileBasedIndexImpl.java:811)
	at com.siberika.idea.pascal.util.ModuleUtil$2.compute(ModuleUtil.java:101)
	at com.siberika.idea.pascal.util.ModuleUtil$2.compute(ModuleUtil.java:97)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:918)
	at com.siberika.idea.pascal.util.ModuleUtil.getCompiledByNameNoCase(ModuleUtil.java:97)
	at com.siberika.idea.pascal.lang.compiled.PPUDecompilerCache.retrieveXml(PPUDecompilerCache.java:111)
	at com.siberika.idea.pascal.lang.compiled.PPUDecompilerCache$Loader.load(PPUDecompilerCache.java:78)
	at com.siberika.idea.pascal.lang.compiled.PPUDecompilerCache$Loader.load(PPUDecompilerCache.java:71)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2276)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
	... 51 more
```

**Comment by casteng** [comment]
Does File -> Invalidate Caches / Restart option help?

**Comment by larrylau** [comment]
> Does File -> Invalidate Caches / Restart option help?

not working

**Comment by boeckmann** [comment]
Same problem here with plugin version 2.40 and the following system:

Intellij IntelliJ IDEA 2019.2.2 (Community Edition)
Build #IC-192.6603.28, built on September 6, 2019
Runtime version: 11.0.3+12-b304.56 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.6
GC: ParNew, ConcurrentMarkSweep
Memory: 989M
Cores: 8
Registry: 
Non-Bundled Plugins: com.siberika.idea.pascal

**Comment by casteng** [comment]
Seems it's affects not only I-Pascal in 2019.x:
https://youtrack.jetbrains.com/issue/IDEA-210623
Will investigate for a workaround.

**Comment by octave-syndicate** [comment]
I can confirm as well.

**Comment by casteng** [comment]
Fixed in v2.50
